Support a "ignore vendor" build_external mode in gazelle and go_repository

Here is a minimal example of a project that cannot build with go_repository:

WORKSPACE:

```python
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "io_bazel_rules_go",
    sha256 = "5756a4ad75b3703eb68249d50e23f5d64eaf1593e886b9aa931aa6e938c4e301",
    urls = ["https://github.com/bazelbuild/rules_go/releases/download/0.14.0/rules_go-0.14.0.tar.gz"],
)

http_archive(
    name = "bazel_gazelle",
    sha256 = "c0a5739d12c6d05b6c1ad56f2200cb0b57c5a70e03ebd2f7b87ce88cabf09c7b",
    urls = ["https://github.com/bazelbuild/bazel-gazelle/releases/download/0.14.0/bazel-gazelle-0.14.0.tar.gz"],
)

load("@io_bazel_rules_go//go:def.bzl", "go_register_toolchains", "go_rules_dependencies")

go_rules_dependencies()

go_register_toolchains()

load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies", "go_repository")

gazelle_dependencies()

go_repository(
    name = "com_github_heroku_docker_registry_client",
    commit = "fd2fe803496897db70c236ee244f41487ea92ec2",
    importpath = "github.com/heroku/docker-registry-client",
)

go_repository(
    name = "com_github_opencontainers_go_digest",
    commit = "279bed98673dd5bef374d3b6e4b09e2af76183bf",
    importpath = "github.com/opencontainers/go-digest",
)
```

BUILD.bazel:

```python
load("@io_bazel_rules_go//go:def.bzl", "go_binary")

go_binary(
    name = "main",
    srcs = [
        "main.go",
    ],
    deps = [
        "@com_github_heroku_docker_registry_client//registry:go_default_library",
        "@com_github_opencontainers_go_digest//:go_default_library",
    ],
)
```

main.go:

```go
package main

import (
	"github.com/heroku/docker-registry-client/registry"
	"github.com/opencontainers/go-digest"
)

func main() {
	reg, _ := registry.New("myregistry.com", "user", "pass")
	layerDigest := digest.FromString("myLayer")
	reg.HasLayer("repo", layerDigest)
}
```
I get the following compile error when running ```bazel build ...``` (Im using bazel version 0.15):

```
main.go:11:15: cannot use layerDigest (type "github.com/opencontainers/go-digest".Digest) as type "github.com/heroku/docker-registry-client/vendor/github.com/opencontainers/go-digest".Digest in argument to reg.HasLayer
GoCompile: error running subcommand: exit status 2
Target //:main failed to build
```

The issue here is that `heroku/docker-registry-client` has a vendored copy of `opencontainers/go-digest`, and it also accepts types from `opencontainers/go-digest` in its public API. I would like gazelle to generate a BUILD file that uses the external version of go-digest (`@com_github_opencontainers_go_digest//:go_default_library`) instead of the vendored.  I've commented on issue #269 believing this to be a bug but after reading [dependency resolution](https://github.com/bazelbuild/bazel-gazelle#dependency-resolution) more carefully I realize this is by design. Gazelle will always find the vendored version at step number 3 regardless of the build_external flag that is passed in via go_repository.

Im proposing we either change build_external to ignore vendor directories if set to "external" or we add a third mode that does so. Im willing to work on the PR if this issue is approved.

--- a/README.rst
+++ b/README.rst
@@ -267,6 +267,16 @@ The following flags are accepted:
 | Bazel may still filter sources with these tags. Use                                              |
 | ``bazel build --features gotags=foo,bar`` to set tags at build time.                             |
 +--------------------------------------------------------------+-----------------------------------+
+| :flag:`-exclude path`                                        |                                   |
++--------------------------------------------------------------+-----------------------------------+
+| Prevents Gazelle from processing a file or directory. If the path refers to                      |
+| a source file, Gazelle won't include it in any rules. If the path refers to                      |
+| a directory, Gazelle won't recurse into it.                                                      |
+|                                                                                                  |
+| This option may be repeated. Paths must be slash-separated, relative to the                      |
+| repository root. This is equivalent to the ``# gazelle:exclude path``                            |
+| directive.                                                                                       |
++--------------------------------------------------------------+-----------------------------------+
 | :flag:`-external external|vendored`                          | :value:`external`                 |
 +--------------------------------------------------------------+-----------------------------------+
 | Determines how Gazelle resolves import paths that cannot be resolve in the                       |

--- a/internal/repos/BUILD.bazel
+++ b/internal/repos/BUILD.bazel
@@ -23,6 +23,7 @@ go_test(
     name = "go_default_test",
     srcs = [
         "import_test.go",
+        "modules_test.go",
         "remote_test.go",
         "repo_test.go",
     ],

--- a/internal/walk/BUILD.bazel
+++ b/internal/walk/BUILD.bazel
@@ -10,6 +10,7 @@ go_library(
     visibility = ["//:__subpackages__"],
     deps = [
         "//internal/config:go_default_library",
+        "//internal/flag:go_default_library",
         "//internal/pathtools:go_default_library",
         "//internal/rule:go_default_library",
     ],

--- a/internal/walk/config.go
+++ b/internal/walk/config.go
@@ -20,6 +20,7 @@ import (
 	"path"
 
 	"github.com/bazelbuild/bazel-gazelle/internal/config"
+	gzflag "github.com/bazelbuild/bazel-gazelle/internal/flag"
 	"github.com/bazelbuild/bazel-gazelle/internal/rule"
 )
 
@@ -48,7 +49,9 @@ func (wc *walkConfig) isExcluded(rel, base string) bool {
 type Configurer struct{}
 
 func (_ *Configurer) RegisterFlags(fs *flag.FlagSet, cmd string, c *config.Config) {
-	c.Exts[walkName] = &walkConfig{}
+	wc := &walkConfig{}
+	c.Exts[walkName] = wc
+	fs.Var(&gzflag.MultiFlag{Values: &wc.excludes}, "exclude", "Path to file or directory that should be ignored (may be repeated)")
 }
 
 func (_ *Configurer) CheckFlags(fs *flag.FlagSet, c *config.Config) error { return nil }

